File: fabric-audit-agent/adapters/store.local.js
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { dirname } from 'node:path';

/**
 * Local-JSON StorePort: persists run history. The prod store (transfer time)
 * implements the same { history, append } against a real DB.
 * @param {string} filePath
 * @param {{ keep?: number }} [options]
 */
export function createLocalStore(filePath, { keep = 180 } = {}) {
  return {
    /** @returns {Promise<Array<{runAt:string, findings:object[]}>>} */
    async history() {
      try {
        return JSON.parse(await readFile(filePath, 'utf-8'));
      } catch (err) {
        if (err.code === 'ENOENT') return [];
        throw err;
      }
    },
    /** @param {{runAt:string, findings:object[]}} run */
    async append(run) {
      const all = await this.history();
      all.push(run);
      const trimmed = all.slice(-keep);
      await mkdir(dirname(filePath), { recursive: true });
      await writeFile(filePath, JSON.stringify(trimmed, null, 2), 'utf-8');
      return trimmed.length;
    },
  };
}

File: fabric-audit-agent-app/fabric_audit_agent/adapters/store_local.py
"""Local-JSON StorePort: persists run history. Port of ``adapters/store.local.js``.

The prod store (Delta/Unity Catalog at deploy) implements the same ``{history, append}``
contract against a real table. ``keep`` trims to the most recent N runs.
"""
import json
import os


def create_local_store(file_path, keep=180):
    def history():
        try:
            with open(file_path, "r", encoding="utf-8") as fh:
                return json.load(fh)
        except FileNotFoundError:
            return []

    def append(run):
        all_runs = history()
        all_runs.append(run)
        trimmed = all_runs[-keep:]
        os.makedirs(os.path.dirname(file_path) or ".", exist_ok=True)
        # Atomic write: a concurrent reader (e.g. the MCP whats_changed tool) must never see a
        # torn/partial file mid-write. Write to a sibling temp file, then os.replace -- atomic
        # on the same filesystem -- so readers always see either the old file or the new one.
        tmp_path = file_path + ".tmp"
        with open(tmp_path, "w", encoding="utf-8") as fh:
            json.dump(trimmed, fh, indent=2, ensure_ascii=False)
        os.replace(tmp_path, file_path)
        return len(trimmed)

    return {"history": history, "append": append}
